README.md
# markdown-img-down-site-change(下载/替换markdown中的图片)
> 搜索目标文件夹中的markdown文件，找到目标图片，提供下载图片，替换图片链接的功能-通常用于markdown 图片失效。

### 简介

这是一个极为轻量的脚本，引用包，设置好参数，通过API即可轻松上手。

### 解决什么问题？

1. 集中下载markdown文件中某个域名下的图片到一个文件夹下。
2. 用新的图片链接替换markdown文件中某个域名的图片链接。

```js
// 1. 下载这两个图片
// ![](https://user-gold-cdn.xitu.io/2019/5/20/图片名字?w=2024&h=1240&f=png&s=339262)
// ![](https://user-gold-cdn.xitu.io/2018/6/16/图片名字)
// 2. 替换成：github的链接
![](https://raw.githubusercontent.com/OBKoro1/articleImg_src/master/juejin/图片名字)
// 由于github的参数限制，带了参数就不能正常显示图片 所有这里默认取图片名前部分的：字母或数字或下划线或汉字
```

### 安装:

```js
npm i markdown-img-down-site-change
```

### 文档：

[Github](https://github.com/OBKoro1/markdown-img-down-site-change)

[API](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/API)

[更新日志](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97)

### 数据安全：

刚上手可能不了解脚本的功能，需要调试一番，这时候万一把`markdown`文件给改坏了，岂不是要哭死？

脚本有两种形式来防止这种情况发生：

1. 脚本会[默认备份](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/API#copy_item_data%E5%A4%87%E4%BB%BD%E9%A1%B9%E7%9B%AE%E7%9A%84%E5%9C%B0%E5%9D%80)你的文件。
2. 默认开启[测试模式](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/API#test%E6%98%AF%E5%90%A6%E5%BC%80%E5%90%AF%E6%B5%8B%E8%AF%95%E6%A8%A1%E5%BC%8F)，等到调试的差不多了，可以关闭测试模式。
3. 建议：再不放心的话，可以先用一两个文件来测试一下脚本

### 插件使用姿势：

我是这样使用的，大家可以做个参考：

1. 新建一个[github项目](https://github.com/OBKoro1/articleImg_src),专门用于存放图片资源
2. `npm init`初始化`package.json` 文件，下载npm包:`npm i markdown-img-down-site-change`,
3. 新建一个`handleImg.js`文件，参考:[example.js](https://github.com/OBKoro1/markdown-img-down-site-change/blob/master/example.js)
4. 拷贝你的`markdown`文件夹到该项目，执行`node handleImg.js`
5. 将会搜索所有`.md`文件，下载匹配到的图片到项目的根目录。
6. 默认不修改文件,默认开启备份，匹配规则、过滤某些文件夹等更多配置信息参阅：[API文档](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/API)

比如我的github图床：[articleImg_src](https://github.com/OBKoro1/articleImg_src)

### 使用：20行代码不到

在项目中有一个使用[栗子](https://github.com/OBKoro1/markdown-img-down-site-change/blob/master/example.js)，里面加了蛮多注释和空行的，实际代码20行都不到，可以说很简单了，如下：

```js
// npm i markdown-img-down-site-change -S 
const markdownImageDown = require('markdown-img-down-site-change'); // 文件模块

// 传参： 这也是脚本的默认参数，根据情况可以自行修改
let option = {
    replace_image_url: 'https://user-gold-cdn.xitu.io/',
    read_markdown_src: './source', // 要查找markdown文件的文件夹地址
    down_img_src: './juejin', // 下载图片到这个文件夹
    var_number: 3 // url前半部分的变量数量 比如上面的日期: /2019/5/20/、/2018/6/16/
}

// 初始化
const markdownImage = new markdownImageDown(option)

// 下载外链
markdownImage.checkDownImg();

// 上传下载下来的图片文件夹到云端 用户自己操作

// 上传图片之后 
// 脚本会把以前的外链替换成云端地址+拼接一个图片名
markdownImage.updateOption({
    new_image_url: 'https://xxx.com/目录地址/', // 图片上传的地址
    add_end: '?raw=true' // github图片地址有后缀 直接进去是仓库
})

// 替换外链 
// 把replace_image_url的字符串换成new_image_url字符串
markdownImage.replaceMarkdown();
```
### 运行：

**仔细阅读文本，配置好参数之后**

在项目根节点新建一个`handleImg.js`文件，安装一下脚本，然后用`node`运行该文件：

```js
npm i markdown-img-down-site-change -S
node handleImg.js
```

### 欢迎试用

有需要的小伙伴，赶紧来试试吧！文档写的很全，上手非常轻松，项目将会持续维护，有什么问题，欢迎给我提[issue](https://github.com/OBKoro1/markdown-img-down-site-change/issues)~

如果觉得这个脚本还不错的话，就给[项目](https://github.com/OBKoro1/markdown-img-down-site-change)点个Star吧！

CHANGELOG.md
### [V1.3.0] 2019/7/9

* 新增选项：[downFileNameCallBack](https://github.com/OBKoro1/markdown-img-down-site-change/wiki/API#downfilenamecallback-%E5%A4%84%E7%90%86%E4%B8%8B%E8%BD%BD%E5%9B%BE%E7%89%87%E7%9A%84name)。
* 下载的图片名变动：
    * 之前下载的图片名字和修改`markdown`链接是将后面部分全都切出来，比如`https://user-gold-cdn.xitu.io/2018/5/1/1631b6f52f7e7015?w=344&h=344&f=jpeg&s=8317`,图片名字是：`1631b6f52f7e7015?w=344&h=344&f=jpeg&s=8317`
    * github上无法显示待参数的图片：`https://github.com/*/1631b6f52f7e7015?w=344&h=344&f=jpeg&s=8317`
    * 现在默认后面的全都不要，正则匹配是：`/^(\w+)/`(取图片名前部分的字母或数字或下划线或汉字),比如该例子下载的图片名为：`1631b6f52f7e7015`
    * 当然下载的图片名变了，修改图片链接时：markdown中的链接也会一起变化的。
    * 如果跟你的需求不符合，也可以使用上面的新增选项`downFileNameCallBack`来自行处理图片名字。

### [V1.1.0] 2019/6/3

* 完善[wiki](https://github.com/OBKoro1/markdown-img-down-site-change/wiki)文档
* 提供下载失效图片的功能
* 提供替换失效图片链接的功能


src/main.js
/*
 * @Github: https://github.com/OBKoro1/markdown-img-down-site-change
 * @Author: OBKoro1
 * @Created_time: 2019-05-28 17:21:41
 * @LastEditors: OBKoro1
 * @LastEditTime: 2019-06-03 15:11:48
 * @Description: 查找指定文件夹的所有markdown文件。
 * 根据参数找出要所有要替换的图片，下载所有图片，替换图片的地址。
 */

const util = require('./util')
const fs = require('fs'); // 文件模块
var request = require('request');
var async = require("async");

class markdownImageDown {
    /**
     * @param {Object} option 配置项
     * new_image_url: '', // 图片上传的地址
     * add_end: '', // 在图片名字后面添加后缀添加后缀
     * read_markdown_src: './source', // 要查找markdown文件的文件夹地址
     * down_img_src: './markdown_img_src', // 下载项目的地址
     * copy_item_data: true, // 默认备份项目 防止项目数据损坏
     * filter_item: [], // 过滤某些文件夹 不去查找markdown
     * new_image_url: '', // 图片上传的地址
     * @Created_time: 2019-05-31 14:30:40
     */
    constructor(option = {}) {
        if (!util.checkDataAction('Object', option, 'option必须为对象')) return
        let defaultOption = {
            replace_image_url: 'http://ww1.sinaimg.cn/large/',
            new_image_url: 'https://xxx.com/目录地址/', // 图片上传的地址
            add_end: '', // 在图片名字后面添加后缀添加后缀
            read_markdown_src: './source', // 要查找markdown文件的文件夹地址
            down_img_src: './markdown_img_src', // 图片下载到这个文件夹下面
            copy_item_data: 'copy_item_obkoro', // 默认备份项目 防止项目数据损坏
            filter_item: ['.git'], // 过滤某些文件夹 不去查找markdown
        }
        this.imgList = [] // 查找的img
        this.option = Object.assign(defaultOption, option) // 配置参数
        if (this.option.copy_item_data) {
            // 异步复制
            this.copyDir(this.option.read_markdown_src, this.option.copy_item_data, (err) => {
                console.log('复制文件夹报错：', err)
            });
        }
    }
    /**
     * 根据类型 开始操作
     * @param {String} type 类型: 'down'只下载图片、'replace' 只替换图片
     * @Created_time: 2019-05-31 14:25:06
     */
    begin(type = 'down') {
        if (!util.checkDataAction('String', type, 'begin type 必须为字符串')) return
        this.type = type

        if (type === 'down') {
            this.readFile(this.option.read_markdown_src)
            console.log('图片提取完成', this.imgList)
            this.downImg()
        }
        // 上传图片到云端 用户自己操作
        if (type === 'replace') {
            this.readFile(this.option.read_markdown_src, true)
            console.log('修改图片地址成功')
        }
    }
    /**
     * 递归查找文件夹，找到markdown文件的图片语法，
     * 匹配要被替换的图片，添加图片到数组/替换图片地址
     * @param {Stying} path 查找的文件夹
     * @param {Bealoon} replace 是否替换查找
     * @return: 
     * @Created_time: 2019-05-29 14:18:28
     */
    readFile(path, replace = false) {
        if (!util.checkDataAction('String', path)) return
        var files = fs.readdirSync(path); // 返回文件数组
        let reg = /!\[(.*)\]\((.*?)\)/g; // 提取markdown图片语法

        files.forEach((item) => {
            let url = `${path}/${item}`; // 文件路径
            let isDirectory = fs.statSync(url).isDirectory(); // 判断是否为文件夹
            if (isDirectory) {
                // 递归文件夹
                if (!this.option.filter_item.includes(url)) {
                    return this.readFile(url, replace)
                }
            } else {
                if (item.indexOf('.md') !== -1) {
                    // 读取文件
                    let data = fs.readFileSync(url, 'utf-8'); // 获取文件内容 返回字符串
                    let res;
                    let isChange = false
                    while ((res = reg.exec(data)) !== null) {
                        let regUrl = res[2]
                        // 添加图片到数组 是否找到该字符串
                        if (regUrl.indexOf(this.option.replace_image_url) !== -1) {
                            if (replace) {
                                // 拼接要替换的url
                                let imgName = regUrl.split(this.option.replace_image_url)[1]
                                let replaceUrl = `${this.option.new_image_url}${imgName}`
                                // 可以添加后缀 如github查看图片后缀为: ?raw=true
                                if (this.option.add_end) {
                                    replaceUrl += this.option.add_end
                                }
                                // 替换字符串
                                isChange = true
                                data = data.replace(regUrl, replaceUrl)
                            } else {
                                this.imgList.push(regUrl)
                            }
                        }
                    }
                    // 修改文件
                    if (replace && isChange) {
                        fs.writeFile(url, data, 'utf-8', () => {
                            console.log('修改成功', url)
                        })
                    }
                }
            }
        })
    }

    // 下载图片
    downImg() {
        // 创建文件夹
        this.mkdirSync();
        async.mapSeries(this.imgList, (httpSrc, callback) => {
            // settimeout等待下载函数创建下载 不需要等下载完毕 是并行
            // 时间充裕的话可以下载函数放进函数中 在回调中调callback
            setTimeout(() => {
                // 图片名+后缀
                let imgName = httpSrc.split(this.option.replace_image_url)
                try {
                    if (imgName[1]) {
                        this.downloadPic(httpSrc, `${this.option.down_img_src}/${imgName[1]}`)
                    }
                    callback(null, httpSrc);
                } catch (err) {
                    // 捕获报错 下载失败
                    callback(err, httpSrc);
                }
            }, 400);
        }, (err, res) => {
            if (err) {
                throw err;
            } else {
                console.log('图片下载完成：', res)
            }
        });
    }
    /**
     * 下载图片 命名冲突会覆盖旧的文件 
     * @param {String} src 图片的下载地址
     * @param {String} imgPath 图片下载到哪里的地址
     * @Created_time: 2019-05-31 14:35:08
     */
    downloadPic(src, imgPath) {
        request(src).pipe(fs.createWriteStream(imgPath)).on('close', () => {
            console.log('pic saved!', src)
        })
    }
    /**
     * 创建文件夹 储存图片的地址
     * @param {type} 
     * @return: 
     * @Created_time: 2019-05-31 14:37:30
     */
    mkdirSync() {
        try {
            let isDirectory = fs.statSync(this.option.down_img_src).isDirectory(); // 判断是否为文件夹
            if (!isDirectory) {
                fs.mkdirSync(this.option.down_img_src, { recursive: true });
            }
        } catch (err) {
            fs.mkdirSync(this.option.down_img_src, { recursive: true });
        }
    }
    /*
     * 复制目录、子目录，及其中的文件
     * @param src {String} 要复制的目录
     * @param dist {String} 复制到目标目录
     */
    copyDir(src, dist, callback) {
        fs.access(dist, function (err) {
            if (err) {
                // 目录不存在时创建目录
                fs.mkdirSync(dist);
            }
            _copy(null, src, dist);
        });

        var _copy = (err, src, dist) => {
            if (err) {
                callback(err);
            } else {
                fs.readdir(src, (err, paths) => {
                    if (err) {
                        callback(err)
                    } else {
                        paths.forEach((path) => {
                            var _src = src + '/' + path;
                            var _dist = dist + '/' + path;
                            fs.stat(_src, (err, stat) => {
                                if (err) {
                                    callback(err);
                                } else {
                                    // 判断是文件还是目录
                                    if (stat.isFile()) {
                                        fs.writeFileSync(_dist, fs.readFileSync(_src));
                                    } else if (stat.isDirectory()) {
                                        // 当是目录是，递归复制
                                        this.copyDir(_src, _dist, callback)
                                    }
                                }
                            })
                        })
                    }
                })
            }
        }
    }
    /**
     * @param {Object} newOption 
     * @return: 
     * @Created_time: 2019-05-31 14:33:00
     */
    updateOption(newOption = {}) {
        if (!util.checkDataAction('Object', newOption, 'updateOption的参数为对象')) return
        this.option = Object.assign(this.option, newOption) // 配置参数
    }

}


module.exports = markdownImageDown


src/index.js
/*
 * @Github: https://github.com/OBKoro1/markdown-img-down-site-change
 * @Author: OBKoro1
 * @Created_time: 2019-05-28 17:21:41
 * @LastEditors: OBKoro1
 * @LastEditTime: 2019-07-11 20:24:14
 * @Description: 查找指定文件夹的所有markdown文件。
 * 根据参数找出要所有要替换的图片，下载所有图片，替换图片的地址。
 */

const util = require('./util')
const fs = require('fs'); // 文件模块
const request = require('request');
const async = require("async");

class markdownImageDown {
    /**
     * @param {Object} option 配置项
     * new_image_url: '', // 图片上传的地址
     * add_end: '', // 在图片名字后面添加后缀添加后缀
     * read_markdown_src: './source', // 要查找markdown文件的文件夹地址
     * down_img_src: './markdown_img_src', // 下载项目的地址
     * copy_item_data: true, // 默认备份项目 防止项目数据损坏
     * filter_item: [], // 过滤某些文件夹 不去查找markdown
     * new_image_url: '', // 图片上传的地址
     * var_number: 0, // url前半部分的变量数量
     * @Created_time: 2019-05-31 14:30:40
     */
    constructor(option = {}) {
        if (!util.checkDataAction('Object', option, 'option必须为对象')) return
        let defaultOption = {
            replace_image_url: 'http://ww1.sinaimg.cn/large/',
            new_image_url: 'https://xxx.com/目录地址/', // 图片上传的地址
            add_end: '', // 在图片名字后面添加后缀添加后缀
            read_markdown_src: './source', // 要查找markdown文件的文件夹地址
            down_img_src: './markdown_img_src', // 图片下载到这个文件夹下面
            copy_item_data: 'copy_item_obkoro', // 默认备份项目 防止项目数据损坏
            var_number: 0, // url前半部分的变量数量
            is_link: true, // 不止匹配![](),链接也匹配 []()
            test: true, // 测试模式：不修改.md
            write_file_time: 3000, // 修改文件的settimeout时间，为拷贝文件留出的时间
            downFileNameCallBack: null, // 处理下载的图片name 返回一个图片name 
            filter_item: ['.git'], // 过滤某些文件夹 不去查找markdown
        }
        this.imgList = [] // 查找的img
        this.option = Object.assign(defaultOption, option) // 配置参数
        if (this.option.copy_item_data) {
            // 异步复制
            this.copyNow = true
            this.copyDir(this.option.read_markdown_src, this.option.copy_item_data, (err) => {
                console.log('复制文件夹报错：', err)
            });
        }
    }
    // 搜索图片链接 下载图片
    checkDownImg() {
        this.readFile(this.option.read_markdown_src)
        console.log(`图片提取完成共：${this.imgList.length}张`, this.imgList)
        this.downImg()
    }
    // 替换图片链接
    replaceMarkdown() {
        let time = 0
        if (this.copyNow) {
            time = this.option.write_file_time
            this.copyNow = false;
        }
        setTimeout(() => {
            this.readFile(this.option.read_markdown_src, true)
            console.log('修改图片地址成功')
        }, time)
    }
    /**
     * @description: 正则匹配整个url或者部分url
     * @param {String} type 'allUrl'：匹配markdown图片
     * @return: reg 
     * @Created_time: 2019-06-04 10:32:50
     */
    reg(type) {
        let replaceSlash = (str) => {
            str = str.replace(/\//g, '\\/')
            str = str.replace(/\./g, '\\.')
            return str
        }
        let jointReg = (num) => {
            let str = ``
            if (num > 0) {
                let defineStr = `.*\\/`
                str += defineStr.repeat(num)
                str += `)`
            } else {
                str = `)` // 直接提取url
            }
            return str
        }
        let http = this.option.replace_image_url
        let url = '!'
        if (this.option.is_link) {
            url = '' // []() 链接形式的图片也匹配
        }
        let regObj = {
            // 根据url和变量匹配url中要切割的部分
            sectionUrl: new RegExp(`(${replaceSlash(http)}${jointReg(this.option.var_number)}`, 'g'),
            // 根据url匹配整个图片url
            allUrl: new RegExp(`${url}\\[(.*)\\]\\((${replaceSlash(this.option.replace_image_url)}.*?)\\)`, 'g')
        }
        return regObj[type]

    }
    /**
     * 递归查找文件夹，找到markdown文件的图片语法，
     * 匹配要被替换的图片，添加图片到数组/替换图片地址
     * @param {Stying} path 查找的文件夹
     * @param {Bealoon} replace 是否替换查找
     * @return: 
     * @Created_time: 2019-05-29 14:18:28
     */
    readFile(path, replace = false) {
        if (!util.checkDataAction('String', path)) return
        var files = fs.readdirSync(path); // 返回文件数组
        let reg = this.reg('allUrl'); // 提取markdown图片语法
        files.forEach((item) => {
            let url = `${path}/${item}`; // 文件路径
            let isDirectory = fs.statSync(url).isDirectory(); // 判断是否为文件夹
            if (isDirectory) {
                // 递归文件夹
                if (!this.option.filter_item.includes(url)) {
                    return this.readFile(url, replace)
                }
            } else {
                if (item.indexOf('.md') !== -1) {
                    // 读取文件
                    let data = fs.readFileSync(url, 'utf-8'); // 获取文件内容 返回字符串
                    let res;
                    let isChange = false

                    while ((res = reg.exec(data)) !== null) {
                        let regUrl = res[2]
                        // 添加图片到数组 是否找到该字符串
                        if (regUrl.indexOf(this.option.replace_image_url) !== -1) {
                            if (replace) {
                                // 做的操作：
                                // http://ww1.sinaimg.cn/large/aaaaa.jpg，![](https://user-gold-cdn.xitu.io/2019/5/20/16ad3ff354f2dac3?w=2024&h=1240&f=png&s=339262)
                                // 变量0：[ '', 'http://ww1.sinaimg.cn/large/', 'aaaaa.jpg' ]
                                // 变量3(日期)：['![](', 'https://user-gold-cdn.xitu.io/2019/5/20/', '16ad3ff354f2dac3?w=2024&h=1240&f=png&s=339262)']
                                try {
                                    let sectionUrlReg = this.reg('sectionUrl')
                                    let splitArr = regUrl.split(sectionUrlReg)
                                    const fileName = this.fileNameFormat(splitArr[2])
                                    let replaceUrl = `${this.option.new_image_url}${fileName}`
                                    // 可以添加后缀 如github查看图片后缀为: ?raw=true
                                    if (this.option.add_end) {
                                        replaceUrl += this.option.add_end
                                    }
                                    // 替换字符串
                                    isChange = true
                                    data = data.replace(regUrl, replaceUrl)
                                } catch (err) {
                                    throw `请检查new_image_url、var_number的设置：${err}`
                                }


                            } else {
                                // 去重
                                if (!this.imgList.includes(regUrl)) {
                                    this.imgList.push(regUrl)
                                }
                            }
                        }
                    }
                    // 修改文件
                    if (replace && isChange) {
                        if (this.option.test) {
                            console.log('测试模式，不修改文件：', url)
                        } else {
                            fs.writeFile(url, data, 'utf-8', () => {
                                console.log('修改成功', url)
                            })
                        }
                    }
                }
            }
        })
    }

    // 下载图片
    downImg() {
        // 创建文件夹
        this.mkdirSync();
        let num = 0
        async.mapSeries(this.imgList, (httpSrc, callback) => {
            // settimeout等待下载函数创建下载 不需要等下载完毕 是并行
            // 时间充裕的话可以下载函数放进函数中 在回调中调callback
            setTimeout(() => {
                // 图片名+后缀
                let sectionUrlReg = this.reg('sectionUrl')
                let splitArr = httpSrc.split(sectionUrlReg)
                try {
                    if (splitArr[2]) {
                        num++
                        const fileName = this.fileNameFormat(splitArr[2])
                        this.downloadPic(httpSrc, `${this.option.down_img_src}/${fileName}`)
                    }
                    callback(null, httpSrc);
                } catch (err) {
                    // 捕获报错 下载失败
                    callback(err, httpSrc);
                }
            }, 400);
        }, (err, res) => {
            if (err) {
                throw err;
            } else {
                console.log('图片下载完成：', res, num)
            }
        });
    }
    /**
     * @description: 处理图片名
     * @param {type} 
     * @return: fileNAme
     * @Date: 2019-07-08 19:51:57
     */
    fileNameFormat(fileNAme) {
        if (this.option.downFileNameCallBack === null) {
            // 默认取图片名前部分的：字母或数字或下划线或汉字
            // 不匹配其他参数符号
            const reg = /^(\w+)/
            const res = fileNAme.match(reg)
            if (res && res[1]) {
                fileNAme = res[1]
            }
        } else {
            if (Object.prototype.toString.call(this.option.downFileNameCallBack) === '[object Function]') {
                fileNAme = this.option.downFileNameCallBack(fileNAme)
                if (typeof fileNAme !== 'string') {
                    throw '处理图片名错误:'
                }
            } else {
                throw '处理图片名错误:'
            }
        }
        return fileNAme
    }
    /**
     * 下载图片 命名冲突会覆盖旧的文件 
     * @param {String} src 图片的下载地址
     * @param {String} imgPath 图片下载到哪里的地址
     * @Created_time: 2019-05-31 14:35:08
     */
    downloadPic(src, imgPath) {
        request(src).pipe(fs.createWriteStream(imgPath)).on('close', () => {
            console.log('pic saved!', src)
        })
    }
    /**
     * 创建文件夹 储存图片的地址
     * @param {type} 
     * @return: 
     * @Created_time: 2019-05-31 14:37:30
     */
    mkdirSync() {
        try {
            let isDirectory = fs.statSync(this.option.down_img_src).isDirectory(); // 判断是否为文件夹
            if (!isDirectory) {
                fs.mkdirSync(this.option.down_img_src, { recursive: true });
            }
        } catch (err) {
            fs.mkdirSync(this.option.down_img_src, { recursive: true });
        }
    }
    /*
     * 复制目录、子目录，及其中的文件
     * @param src {String} 要复制的目录
     * @param dist {String} 复制到目标目录
     */
    copyDir(src, dist, callback) {
        fs.access(dist, (err) => {
            if (err) {
                // 目录不存在时创建目录
                fs.mkdir(dist, { recursive: true }, (err) => {
                    if (err) throw err;
                    _copy(null, src, dist);
                });
            } else {
                this.copyNow = false
                return
            }
        });

        var _copy = (err, src, dist) => {
            if (err) {
                callback(err);
            } else {
                fs.readdir(src, (err, paths) => {
                    if (err) {
                        callback(err)
                    } else {
                        paths.forEach((path) => {
                            var _src = src + '/' + path;
                            var _dist = dist + '/' + path;
                            fs.stat(_src, (err, stat) => {
                                if (err) {
                                    callback(err);
                                } else {
                                    // 判断是文件还是目录
                                    if (stat.isFile()) {
                                        fs.readFile(_src, (err, data) => {
                                            if (err) throw err;
                                            fs.writeFile(_dist, data, (err) => {
                                                if (err) {
                                                    throw `文件拷贝失败：${err}`;
                                                }
                                                console.log('拷贝', _dist)
                                            });
                                        })
                                    } else if (stat.isDirectory()) {
                                        // 当是目录是，递归复制
                                        this.copyDir(_src, _dist, callback)
                                    }
                                }
                            })
                        })
                    }
                })
            }
        }
    }
    /**
     * @param {Object} newOption 
     * @return: 
     * @Created_time: 2019-05-31 14:33:00
     */
    updateOption(newOption = {}) {
        if (!util.checkDataAction('Object', newOption, 'updateOption的参数为对象')) return
        this.option = Object.assign(this.option, newOption) // 配置参数
    }

}


module.exports = markdownImageDown


example.js
/*
 * @Github: https://github.com/OBKoro1/markdown-img-down-site-change
 * @Author: OBKoro1
 * @Created_time: 2019-05-31 16:05:19
 * @LastEditors: OBKoro1
 * @LastEditTime: 2019-06-03 11:09:33
 * @Description: 查找指定文件夹的所有markdown文件。
 * 根据参数找出要所有要替换的图片，下载所有图片，替换图片的地址。
 */
// npm i markdown-img-down-site-change -S 
const markdownImageDown = require('markdown-img-down-site-change'); // 文件模块

// 传参： 这也是脚本的默认参数，根据情况可以自行修改
let option = {
    replace_image_url: 'http://ww1.sinaimg.cn/large/', // 要被替换的图片
    read_markdown_src: './source', // 要查找markdown文件的文件夹地址
    down_img_src: './markdown_img_src', // 下载图片到这个文件夹
    filter_item: ['.git'], // 过滤某些文件夹 不去查找markdown
}

// 初始化
const markdownImage = new markdownImageDown(option)

// 下载外链
markdownImage.begin('down');

// 上传下载下来的图片文件夹到云端 用户自己操作

// 上传图片之后 
// 脚本会把以前的外链替换成云端地址+拼接一个图片名
markdownImage.updateOption({
    new_image_url: 'https://xxx.com/目录地址/', // 图片上传的地址
    add_end: '?raw=true' // github图片地址有后缀 直接进去是仓库
})

// 替换外链 
// 把replace_image_url的字符串换成new_image_url字符串
markdownImage.begin('replace');